### PR TITLE
[Metal][ops] propagate NaN in argmax/argmin to match CPU semantics (#130295)

### LINKED ISSUE #130295: BUG: max, argmax NaN handling with MPS backend vs CPU backend
I don't see any related MPS `max` reproducers quite this simple on the issue tracker, so figured this might help. Working on an M3 Max, Python `3.12` with `torch` `2.3.1`, I saw the `max` discrepancy vs. CPU below. This came up while I was investigating array API conformance in SciPy. I wonder if other more complex neural network MPS issues like https://github.com/pytorch/pytorch/issues/122030 might ultimately reduce to problems similar to this given that `argmax` is also similarly affected and obviously heavily used.

I was initially thinking this behavior might be "ok" because of https://github.com/pytorch/pytorch/issues/116567#issuecomment-1917924375, where it is clarified that NaN sort-order is unspecified in `torch`. However, the array API specifies a requirement for NaN propagation from `max` (https://data-apis.org/array-api/latest/API_specification/generated/array_api.max.html#max) and the `cpu` device is already conformant.

Running:

```python
import torch 

print(torch.__version__)
for device_name in ["cpu", "mps"]:
    tensor = torch.tensor([1.0, float("nan"), 3.0], device=torch.device(device_name))
    max = torch.max(tensor)
    print(f"{device_name=}, {max=}")
```

Produces:

```
2.3.1
device_name='cpu', max=tensor(nan)
device_name='mps', max=tensor(3., device='mps:0')
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

### PR DESCRIPTION
**Pull request by versi379**
## Problem

`torch.argmax` / `torch.argmin` on MPS ignore `NaN` values. CPU returns the index of the first `NaN` (NaN propagates through max/min); MPS returns the index of the largest finite value. The `max` / `min` half of #130295 was fixed in #130445; this PR completes #130295 by handling `argmax` / `argmin`.

## Reproducer (from the issue)

```python
import torch
torch.manual_seed(0)
x = torch.rand(32, device="mps"); x[5] = torch.nan
print(x.argmin(), x.argmax())
```

| | before | after | CPU |
|---|---|---|---|
| `argmin` | `tensor(13)` | `tensor(5)` | `tensor(5)` |
| `argmax` | `tensor(3)`  | `tensor(5)` | `tensor(5)` |

## How

In `argmax_argmin_out_mps`, when input is floating-point: build a 0/1 NaN mask, run argmax on the mask to get the first-NaN index, and `select` between that index and the regular argmax result based on whether any NaN was present. Stays in the MPSGraph path; the added reductions over the NaN-mask are dominated by the existing argmax reduction in typical workloads.

A shader-based version (folding NaN detection and argmax into a single pass) remains a viable follow-up if performance ever motivates it.

## Test

```
python test/test_mps_issue_130295.py -v
```

5 cases pass on Apple M4 Pro: scalar with NaN, per-dim with NaN, no-NaN regression, all-NaN, fp16 with NaN.

Happy to relocate the test into `test/test_mps.py` if preferred.

Fixes #130295.

### CODE REVIEW

File: test/test_mps_issue_130295.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+argmax/argmin ignore NaN on MPS. CPU returns the index of the first NaN
+(NaN propagates through max/min).
+"""
+
+import torch
+from torch.testing._internal.common_utils import run_tests, TestCase
+
+
+class TestArgmaxArgminNaN(TestCase):
+    def test_argmax_with_nan(self):
```
Comment (kurtamohler) [inline test/test_mps_issue_130295.py]: Could you move these tests to the `TestMPS` class in `test_mps.py`? Right now CI won't pick these up

File: aten/src/ATen/native/mps/operations/ReduceOps.mm [hunk 2]
```diff
@@ -859,6 +859,20 @@ static void argmax_argmin_out_mps(const Tensor& input_t,
         argreduceOutTensor = [mpsGraph reductionArgMinimumWithTensor:castInputTensor axis:(NSInteger)dim_ name:nil];
       }
 
+      // MPSGraph's argmax/argmin ignore NaN. CPU returns the index of the first NaN
+      // (NaN propagates through max/min). Detect NaNs and override the result.
+      if (isFloatingType(inputScalarType)) {
+        MPSGraphTensor* nanMask = [mpsGraph isNaNWithTensor:castInputTensor name:@"is_nan"];
+        MPSGraphTensor* nanCast = [mpsGraph castTensor:nanMask toType:MPSDataTypeInt32 name:@"nan_to_int"];
+        MPSGraphTensor* nanIdx = [mpsGraph reductionArgMaximumWithTensor:nanCast axis:(NSInteger)dim_ name:@"nan_idx"];
+        MPSGraphTensor* hasNan = [mpsGraph reductionMaximumWithTensor:nanCast axis:(NSInteger)dim_ name:@"has_nan"];
```
Comment (kurtamohler) [inline aten/src/ATen/native/mps/operations/ReduceOps.mm]: Isn't it possible to avoid this second reduction by just doing `hasNan = nanMask[nanIdx]`?

File: test/test_mps_issue_130295.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        x_cpu = x.cpu()
+        self.assertEqual(torch.argmax(x).item(), torch.argmax(x_cpu).item())
+        self.assertEqual(torch.argmin(x).item(), torch.argmin(x_cpu).item())
+
+    def test_argmax_all_nan(self):
+        x = torch.tensor([float("nan"), float("nan"), float("nan")], device="mps")
+        x_cpu = x.cpu()
+        self.assertEqual(torch.argmax(x).item(), torch.argmax(x_cpu).item())
+
+    def test_argmax_half_nan(self):
```
Comment (kurtamohler) [inline test/test_mps_issue_130295.py]: Rather than write a whole other test for half, you can just decorate the others with `@parametrize("dtype", [torch.float16, torch.bfloat16, torch.float])` (might as well do bfloat16 as well).

File: test/test_mps_issue_130295.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            [[1.0, float("nan"), 3.0], [float("nan"), 2.0, 1.0], [4.0, 5.0, 6.0]],
+            device="mps",
+        )
+        x_cpu = x.cpu()
+        self.assertEqual(torch.argmax(x, dim=0).cpu(), torch.argmax(x_cpu, dim=0))
+        self.assertEqual(torch.argmax(x, dim=1).cpu(), torch.argmax(x_cpu, dim=1))
+        self.assertEqual(torch.argmin(x, dim=0).cpu(), torch.argmin(x_cpu, dim=0))
+        self.assertEqual(torch.argmin(x, dim=1).cpu(), torch.argmin(x_cpu, dim=1))
+
+    def test_argmax_no_nan(self):
```
Comment (kurtamohler) [inline test/test_mps_issue_130295.py]: This test isn't necessary, we already have existing tests for the no-nan case

**Comment by versi379** [comment]
@kurtamohler thanks for the review :)
I have pushed b4cd80c with:

  - Tests moved into `test/test_mps.py` as `TestMPS.test_argmax_argmin_with_nan`
  - `gatherAlongAxis` replaces the second reduction (`hasNan = nanMask[nanIdx]`)
  - Single `@parametrize` test covers fp16, bf16, fp32
  - Dropped the redundant no-nan case

**Comment by kurtamohler** [comment]
Thank you for the PR, but it looks like argmax and argmin are going to be ported to a Metal kernel that will solve this problem (#187304)
